File: JavaScript/12_IntegerToRoman.js
/**
 * @param {number} num
 * @return {string}
 */
const intToRoman = function(num) {
    const letterValues =  new Map([
        ["M", 1000],
        ["CM", 900],
        ["D", 500],
        ["CD", 400],
        ["C", 100],
        ["XC", 90],
        ["L", 50],
        ["XL", 40],
        ["X", 10],
        ["IX", 9],
        ["V", 5],
        ["IV", 4],
        ["I", 1]
    ]);
    let roman = "";

    for (const [key, value] of letterValues) {
        while ((num / value) >= 1) {
            roman += key;
            num -= value;
        }
    }
    return roman;
};

const testCases = [
    {
        num: 3749,
        expected: "MMMDCCXLIX"
    },
    {
        num: 1469,
        expected: "MCDLXIX"
    },
    {
        num: 58,
        expected: "LVIII"
    },
    {
        num: 1994,
        expected: "MCMXCIV"
    },
]

testCases.forEach((testCase, i) => {
    const output = intToRoman(testCase.num);
    if (JSON.stringify(output) === JSON.stringify(testCase.expected)) {
        console.log(`Test Case ${i}: Passed. Expected: ${JSON.stringify(testCase.expected)}, Output: ${JSON.stringify(output)} `)
    } else {
        console.log(`Test Case ${i}: Failed. Expected: ${JSON.stringify(testCase.expected)}, Output: ${JSON.stringify(output)} `)
    }
})

File: CSharp/12_IntegerToRoman.cs
namespace LeetCodeSolutions {
    public static class IntegerToRoman {
        public static string Solution(int num) {
            Dictionary<string, int> letterValues = new Dictionary<string, int> {
                {"M", 1000},
                {"CM", 900},
                {"D", 500},
                {"CD", 400},
                {"C", 100},
                {"XC", 90},
                {"L", 50},
                {"XL", 40},
                {"X", 10},
                {"IX", 9},
                {"V", 5},
                {"IV", 4},
                {"I", 1}
            };
            string roman = "";

            foreach (var lV in letterValues) {
                while (num >= lV.Value) {
                    roman += lV.Key;
                    num -= lV.Value;
                }
            }
            return roman;
        }

        public static void RunTests() {
            Console.WriteLine("Running Integer To Roman Tests...");

            var testCases = new List<(int, string)> {
                (3749, "MMMDCCXLIX"),
                (1469, "MCDLXIX"),
                (58, "LVIII"),
                (1994, "MCMXCIV")
            };

            foreach (var (num, expected) in testCases) {
                try {
                    var result = Solution(num);
                    if (result == expected) {
                        Console.WriteLine(
                            $"Test Passed: Input: {num}, Expected: {expected}, Result: {result}"
                        );
                    } else {
                        Console.WriteLine(
                            $"Test Failed: Input: {num}, Expected: {expected}, Result: {result}"
                        );
                    } 
                } catch (Exception ex) {
                    Console.WriteLine($"Test Failed: {ex.Message}");
                }
            }
        }      
    }
}